heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,270 @@
+  descriptor?: {
+    id: string;
+  };
+  suppressions?: ISuppressedAnalysis[];
+}
+
+interface ISuppressedAnalysis {
+  kind: string;
+  justification: string;
+}
+
+// Interface for the SARIF log structure
+interface ISarifLog {
+  version: string;
+  $schema: string;
+  runs: ISarifRun[];
+}
+
+interface IRegion {
+  startLine?: number;
+  startColumn?: number;
+  endLine?: number;
+  endColumn?: number;
+  snippet?: {
+    text: string;
+  };
+}
+
+interface IStaticAnalysisRules {
+  id: string;
+  name?: string;
+  shortDescription?: {
+    text: string;
+  };
+  fullDescription?: {
+    text: string;
+  };
+  defaultConfiguration?: {
+    level: 'note' | 'warning' | 'error';
+  };
+  helpUri?: string;
+  properties?: {
+    category?: string;
+    precision?: 'very-high' | 'high' | 'medium' | 'low';
+    tags?: string[];
+    problem?: {
+      severity?: 'recommendation' | 'warning' | 'error';
+      securitySeverity?: number;
+    };
+  };
+}
+
+interface ISarifLocation {
+  physicalLocation: ISarifPhysicalLocation;
+}
+
+interface ISarifPhysicalLocation {
+  artifactLocation: {
+    uri: string;
+    index: number;
+  };
+  region?: IRegion;
+}
+
+interface IMessage extends TEslint.Linter.LintMessage {
+  suppressions?: ISuppressedAnalysis[];
+}
+
+interface IExtendedLintResult extends TEslint.ESLint.LintResult {
+  suppressedMessages: TEslint.ESLint.LintResult['suppressedMessages'];
+}
+
+const sarifFiles: Map<string, ISarifFile> = new Map();
+const sarifResults: ISarifRepresentation[] = [];
+
+const internalErrorId: string = 'ESL0999';
+const toolConfigurationNotifications: ISarifRepresentation[] = [];
+
+/**
+ * Converts ESLint results into a SARIF (Static Analysis Results Interchange Format) log.
+ *
+ * This function takes in a list of ESLint lint results, processes them to extract
+ * relevant information such as errors, warnings, and suppressed messages, and
+ * outputs a SARIF log which conforms to the SARIF v2.1.0 specification.
+ *
+ * @param results - An array of lint results (`IExtendedLintResult[]`) from ESLint that contains linting information,
+ *                  such as file paths, messages, and suppression details.
+ * @param options - An object (`ISerifFormatterOptions`) containing options for formatting:
+ *                  - `ignoreSuppressed`: Boolean flag to decide whether to ignore suppressed messages.
+ *                  - `eslintVersion`: Optional string to include the version of ESLint in the SARIF log.
+ * @returns The SARIF log (`ISarifLog`) containing information about the linting results in SARIF format.
+ */
+export function formatAsSARIF(results: IExtendedLintResult[], options: ISerifFormatterOptions): ISarifLog {
+  const { ignoreSuppressed, eslintVersion } = options;
+  const sarifRun: ISarifRun = {
+    tool: {
+      driver: {
+        name: 'ESLint',
+        informationUri: 'https://eslint.org',
+        rules: []
+      }
+    }
+  };
+
+  const sarifLog: ISarifLog = {
+    version: '2.1.0',
```

Comment:
Where does this version come from? Is there a way to get this from the SARIF package?

Reply:
The version is the current SARIF version. There isn't a simple way to fetch the version without using node-fetch. The version also changes really infrequently. It's been at version 2,1,0 since 2019.